Bisect FX node asserts on `ValidationException`.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107493
* #107989

This PR introduces binary search for finding smaller validation errors, when they occur.

We do that by bisecting the sequence of `torch._assert` FX nodes recorded as the source
expression of the translation validator (TV) by `ShapeEnv.evaluate_expr` calls. Then, we
raise the error caused by the earliest node.

In summary, the changes are:
- Call `bisect` on `ValidationError` @ _torch/_dynamo/convert_frame.py_
- Implement the binary search @ _torch/fx/experimental/symbolic_shapes.py_

Edit: moved `ShapeEnv` replay-recording to #107989

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @anijain2305 @ipiszy

ysiraichi committed:
Bisect FX node asserts on `ValidationException`.

This PR introduces binary search for finding smaller validation errors, when they occur.

We do that by bisecting the sequence of `torch._assert` FX nodes recorded as the source
expression of the translation validator (TV) by `ShapeEnv.evaluate_expr` calls. Then, we
raise the error caused by the earliest node.

In summary, the changes are:
- Call `bisect` on `ValidationError` @ _torch/_dynamo/convert_frame.py_
- Store a reference of the list of `TrackedFake` to `ShapeEnv` @ _torch/_dynamo/output_graph.py_
- Implement the event recording logic @ _torch/fx/experimental/symbolic_shapes.py_
    - Create the `ShapeEnvEvent` class
    - Decorate `ShapeEnv` methods with the `record_shapeenv_event` decorator
- Implement the `ShapeEnv` reconstruction function @ _torch/fx/experimental/symbolic_shapes.py_
- Implement the binary search @ _torch/fx/experimental/symbolic_shapes.py_

[ghstack-poisoned]

ysiraichi committed:
Fix lint issues. on "Bisect FX node asserts on `ValidationException`."


This PR introduces binary search for finding smaller validation errors, when they occur.

We do that by bisecting the sequence of `torch._assert` FX nodes recorded as the source
expression of the translation validator (TV) by `ShapeEnv.evaluate_expr` calls. Then, we
raise the error caused by the earliest node.

In summary, the changes are:
- Call `bisect` on `ValidationError` @ _torch/_dynamo/convert_frame.py_
- Store a reference of the list of `TrackedFake` to `ShapeEnv` @ _torch/_dynamo/output_graph.py_
- Implement the event recording logic @ _torch/fx/experimental/symbolic_shapes.py_
    - Create the `ShapeEnvEvent` class
    - Decorate `ShapeEnv` methods with the `record_shapeenv_event` decorator
- Implement the `ShapeEnv` reconstruction function @ _torch/fx/experimental/symbolic_shapes.py_
- Implement the binary search @ _torch/fx/experimental/symbolic_shapes.py_

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ysiraichi committed:
Revert deps change. on "Bisect FX node asserts on `ValidationException`."


This PR introduces binary search for finding smaller validation errors, when they occur.

We do that by bisecting the sequence of `torch._assert` FX nodes recorded as the source
expression of the translation validator (TV) by `ShapeEnv.evaluate_expr` calls. Then, we
raise the error caused by the earliest node.

In summary, the changes are:
- Call `bisect` on `ValidationError` @ _torch/_dynamo/convert_frame.py_
- Store a reference of the list of `TrackedFake` to `ShapeEnv` @ _torch/_dynamo/output_graph.py_
- Implement the event recording logic @ _torch/fx/experimental/symbolic_shapes.py_
    - Create the `ShapeEnvEvent` class
    - Decorate `ShapeEnv` methods with the `record_shapeenv_event` decorator
- Implement the `ShapeEnv` reconstruction function @ _torch/fx/experimental/symbolic_shapes.py_
- Implement the binary search @ _torch/fx/experimental/symbolic_shapes.py_

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ysiraichi committed:
Update on "Bisect FX node asserts on `ValidationException`."


This PR introduces binary search for finding smaller validation errors, when they occur.

We do that by bisecting the sequence of `torch._assert` FX nodes recorded as the source
expression of the translation validator (TV) by `ShapeEnv.evaluate_expr` calls. Then, we
raise the error caused by the earliest node.

In summary, the changes are:
- Call `bisect` on `ValidationError` @ _torch/_dynamo/convert_frame.py_
- Store a reference of the list of `TrackedFake` to `ShapeEnv` @ _torch/_dynamo/output_graph.py_
- Implement the event recording logic @ _torch/fx/experimental/symbolic_shapes.py_
    - Create the `ShapeEnvEvent` class
    - Decorate `ShapeEnv` methods with the `record_shapeenv_event` decorator
- Implement the `ShapeEnv` reconstruction function @ _torch/fx/experimental/symbolic_shapes.py_
- Implement the binary search @ _torch/fx/experimental/symbolic_shapes.py_

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ysiraichi committed:
Update on "Bisect FX node asserts on `ValidationException`."


This PR introduces binary search for finding smaller validation errors, when they occur.

We do that by bisecting the sequence of `torch._assert` FX nodes recorded as the source
expression of the translation validator (TV) by `ShapeEnv.evaluate_expr` calls. Then, we
raise the error caused by the earliest node.

In summary, the changes are:
- Call `bisect` on `ValidationError` @ _torch/_dynamo/convert_frame.py_
- Store a reference of the list of `TrackedFake` to `ShapeEnv` @ _torch/_dynamo/output_graph.py_
- Implement the event recording logic @ _torch/fx/experimental/symbolic_shapes.py_
    - Create the `ShapeEnvEvent` class
    - Decorate `ShapeEnv` methods with the `record_shapeenv_event` decorator
- Implement the `ShapeEnv` reconstruction function @ _torch/fx/experimental/symbolic_shapes.py_
- Implement the binary search @ _torch/fx/experimental/symbolic_shapes.py_

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ysiraichi committed:
Update on "Bisect FX node asserts on `ValidationException`."


This PR introduces binary search for finding smaller validation errors, when they occur.

We do that by bisecting the sequence of `torch._assert` FX nodes recorded as the source
expression of the translation validator (TV) by `ShapeEnv.evaluate_expr` calls. Then, we
raise the error caused by the earliest node.

In summary, the changes are:
- Call `bisect` on `ValidationError` @ _torch/_dynamo/convert_frame.py_
- Store a reference of the list of `TrackedFake` to `ShapeEnv` @ _torch/_dynamo/output_graph.py_
- Implement the event recording logic @ _torch/fx/experimental/symbolic_shapes.py_
    - Create the `ShapeEnvEvent` class
    - Decorate `ShapeEnv` methods with the `record_shapeenv_event` decorator
- Implement the `ShapeEnv` reconstruction function @ _torch/fx/experimental/symbolic_shapes.py_
- Implement the binary search @ _torch/fx/experimental/symbolic_shapes.py_

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ysiraichi committed:
Fix and add tests. on "Bisect FX node asserts on `ValidationException`."


This PR introduces binary search for finding smaller validation errors, when they occur.

We do that by bisecting the sequence of `torch._assert` FX nodes recorded as the source
expression of the translation validator (TV) by `ShapeEnv.evaluate_expr` calls. Then, we
raise the error caused by the earliest node.

In summary, the changes are:
- Call `bisect` on `ValidationError` @ _torch/_dynamo/convert_frame.py_
- Store a reference of the list of `TrackedFake` to `ShapeEnv` @ _torch/_dynamo/output_graph.py_
- Implement the event recording logic @ _torch/fx/experimental/symbolic_shapes.py_
    - Create the `ShapeEnvEvent` class
    - Decorate `ShapeEnv` methods with the `record_shapeenv_event` decorator
- Implement the `ShapeEnv` reconstruction function @ _torch/fx/experimental/symbolic_shapes.py_
- Implement the binary search @ _torch/fx/experimental/symbolic_shapes.py_

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ysiraichi committed:
Update on "Bisect FX node asserts on `ValidationException`."


This PR introduces binary search for finding smaller validation errors, when they occur.

We do that by bisecting the sequence of `torch._assert` FX nodes recorded as the source
expression of the translation validator (TV) by `ShapeEnv.evaluate_expr` calls. Then, we
raise the error caused by the earliest node.

In summary, the changes are:
- Call `bisect` on `ValidationError` @ _torch/_dynamo/convert_frame.py_
- Store a reference of the list of `TrackedFake` to `ShapeEnv` @ _torch/_dynamo/output_graph.py_
- Implement the event recording logic @ _torch/fx/experimental/symbolic_shapes.py_
    - Create the `ShapeEnvEvent` class
    - Decorate `ShapeEnv` methods with the `record_shapeenv_event` decorator
- Implement the `ShapeEnv` reconstruction function @ _torch/fx/experimental/symbolic_shapes.py_
- Implement the binary search @ _torch/fx/experimental/symbolic_shapes.py_

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ezyang commented on test/dynamo/test_exc.py:
```diff
@@ -195,47 +195,132 @@ def fn001(x):
+            ValidationException,
+            lambda: fn(torch.randn(20), (5, 10, 5)),
+            """\
+translation validation failed.
```
How exactly am I supposed to interpret the message here?

ysiraichi replied:
The problem that TV tries to solve is: _"Given **Assertions**, is there any **Model** that satisfies **Target Expressions**, but doesn't satisfy the source expressions?"_. The **Faiiled Source Expressions** represents an affirmative response, showing which source expressions that failed.

ysiraichi replied:
Oh, I see. I thought the smaller problem instance would be a nice win already, e.g. easier to read and reason about. But that makes sense.

ysiraichi replied:
Another thing I thought was that it would be nice to know under what program FX node we failed. But, is there any way to get this information at this point (inside `ShapeEnv`)?

ezyang replied:
Yeah, the stack trace on the bad guard would be pretty useful.

ysiraichi replied:
What do you think of this?

ezyang replied:
Yes, this is better.

ezyang commented on torch/_dynamo/output_graph.py:
```diff
@@ -263,6 +269,7 @@ def __init__(
                 allow_dynamic_output_shape_ops=config.capture_dynamic_output_shape_ops,
                 frame_id=frame_state["_id"],
                 co_fields=self.co_fields,
+                tracked_fakes=self.tracked_fakes,
```
Mutable list sharing here is very delicate, call it out (and make sure it is tested!)

ysiraichi replied:
One alternative I can think of is: 

- New `ShapeEnv.tracked_fakes_length` field
- New `ShapeEnv.inc_tracked_fakes_length` function
    - Call it whenever `tracked_fakes` is modified

In order to make it less error-prone, what do you think of creating a class for automating that?

```python
class TrackedFakeList(list):
    def __init__(self, shape_env):
        self.shape_env = shape_env
        super().__init__()

    def append(self, obj):
        self.shape_env.inc_tracked_fakes_length()
        super().append(obj)
```

ezyang replied:
I don't have a good alternative, what you have right now is OK, it's just delicate.

ysiraichi replied:
Would you rather have this circular reference? I kind of like this alternative idea...

ezyang replied:
Alternate idea is fine too.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1976,6 +1976,119 @@ def print_results(grouped, indent, result_fn):
+            replacearg(index=3, key="fx_node", fn=convert_node)
+
+        # Actually call the method with the converted arguments.
+        return self.f(shape_env, *args, **kwargs)
```
This seems pretty delicate; in particular, it seems difficult to guarantee that the replay mechanism is going to keep working as people keep monkeying around with methods on ShapeEnv. First off, is this well tested? Second, one way to make it structurally harder to "mess it up" is to force all substantive method calls through this API, so that even regular, non bisecting ShapeEnv calls go through the logic here.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1976,6 +1976,119 @@ def print_results(grouped, indent, result_fn):
+# remember_fakes_length: flags whether we should add the length of the
+# ShapeEnv.tracked_fakes list to the event. This is used for calling
+# ShapeEnv.produce_guards at an arbitrary time (assuming the fakes list only
+# increases monotonically).
```
If I add a new method which mutates the state of ShapeEnv, but I didn't read this code carefully enough, how will I know to apply this decorator? What test will tell me I did it wrong?

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2108,15 +2223,36 @@ def __init__(
+                })
+            ]
+
+            self.tracked_fakes = tracked_fakes
```
Note: this will almost certainly cause reference cycles, as tracked fakes has FakeTensor which has reference to SymInt which refer to ShapeEnv

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2108,15 +2223,36 @@ def __init__(
+                    "duck_shape": duck_shape,
+                    "frame_id": frame_id,
+                    "co_fields": co_fields,
+                    "should_record_events": False,
```
If I add new arguments to the constructor, how do I know to add them here?

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -3400,6 +3580,10 @@ def evaluate_expr(self, orig_expr: "sympy.Expr", hint=None, fx_node=None):
 
+            assert node is not None
+            if fresh:
+                node.meta["event"] = len(self.events) - 1
```
what's this?

ysiraichi replied:
This is so we know, when bisecting, what event corresponds to a given node. Say we want to replay until FX node `n`. This tells us until what event index we should replay to get to `n`.

ezyang commented:
I'm not convinced about the ongoing maintainability of the record-replay mechanism on top of ShapeEnv. It feels like it needs some preparatory refactors to justify itself. Try to convince me?

ysiraichi commented:
Thank you for the review! You are right and your comments are on point. I will address all of them and try to make this PR more robust. I will ping you when I get it ready.

ysiraichi commented:
@ezyang I'm thinking about splitting this PR into 2:
1. Introduce replay mechanism for `ShapeEnv`
2. Bisect FX node asserts on `ValidationException`

I think that would be easier to review.
What do you think?

ezyang commented:
That's fine, do as many splits as is not annoying for you to do.

ysiraichi committed:
Update on "Bisect FX node asserts on `ValidationException`."


This PR introduces binary search for finding smaller validation errors, when they occur.

We do that by bisecting the sequence of `torch._assert` FX nodes recorded as the source
expression of the translation validator (TV) by `ShapeEnv.evaluate_expr` calls. Then, we
raise the error caused by the earliest node.

In summary, the changes are:
- Call `bisect` on `ValidationError` @ _torch/_dynamo/convert_frame.py_
- Store a reference of the list of `TrackedFake` to `ShapeEnv` @ _torch/_dynamo/output_graph.py_
- Implement the event recording logic @ _torch/fx/experimental/symbolic_shapes.py_
    - Create the `ShapeEnvEvent` class
    - Decorate `ShapeEnv` methods with the `record_shapeenv_event` decorator
- Implement the `ShapeEnv` reconstruction function @ _torch/fx/experimental/symbolic_shapes.py_
- Implement the binary search @ _torch/fx/experimental/symbolic_shapes.py_

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ysiraichi committed:
Update on "Bisect FX node asserts on `ValidationException`."


This PR introduces binary search for finding smaller validation errors, when they occur.

We do that by bisecting the sequence of `torch._assert` FX nodes recorded as the source
expression of the translation validator (TV) by `ShapeEnv.evaluate_expr` calls. Then, we
raise the error caused by the earliest node.

In summary, the changes are:
- Call `bisect` on `ValidationError` @ _torch/_dynamo/convert_frame.py_
- Implement the binary search @ _torch/fx/experimental/symbolic_shapes.py_

Edit: moved `ShapeEnv` replay-recording to #107989

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ysiraichi committed:
Update on "Bisect FX node asserts on `ValidationException`."


This PR introduces binary search for finding smaller validation errors, when they occur.

We do that by bisecting the sequence of `torch._assert` FX nodes recorded as the source
expression of the translation validator (TV) by `ShapeEnv.evaluate_expr` calls. Then, we
raise the error caused by the earliest node.

In summary, the changes are:
- Call `bisect` on `ValidationError` @ _torch/_dynamo/convert_frame.py_
- Implement the binary search @ _torch/fx/experimental/symbolic_shapes.py_

Edit: moved `ShapeEnv` replay-recording to #107989

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ysiraichi committed:
Fix lint issues. on "Bisect FX node asserts on `ValidationException`."


This PR introduces binary search for finding smaller validation errors, when they occur.

We do that by bisecting the sequence of `torch._assert` FX nodes recorded as the source
expression of the translation validator (TV) by `ShapeEnv.evaluate_expr` calls. Then, we
raise the error caused by the earliest node.

In summary, the changes are:
- Call `bisect` on `ValidationError` @ _torch/_dynamo/convert_frame.py_
- Implement the binary search @ _torch/fx/experimental/symbolic_shapes.py_

Edit: moved `ShapeEnv` replay-recording to #107989

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ysiraichi commented:
@ezyang I finished refactoring this PR. I believe all tests should be passing now. Could you take a look at it again?
Here is a short summary of the changes:

- Record/Replay of `ShapeEnv` was moved to #107989
    - Created a method for comparing the state of a `ShapeEnv` instance
    - Added more comments about events recording and `ShapeEnv` state equality check
    - Added a dynamo config for checking if we recorded the events correctly in the beginning of `produce_guards`
    - Off by default, but on when running _test/dynamo/test_dynamic_shapes.py_
    - Main execution path goes through events `run` (instead of a `__call__` overload) method
- Keeping track of the currently executed FX node with a context manager: `set_current_node`
- Improved `BisectValidationException` message

ysiraichi committed:
Update on "Bisect FX node asserts on `ValidationException`."


This PR introduces binary search for finding smaller validation errors, when they occur.

We do that by bisecting the sequence of `torch._assert` FX nodes recorded as the source
expression of the translation validator (TV) by `ShapeEnv.evaluate_expr` calls. Then, we
raise the error caused by the earliest node.

In summary, the changes are:
- Call `bisect` on `ValidationError` @ _torch/_dynamo/convert_frame.py_
- Implement the binary search @ _torch/fx/experimental/symbolic_shapes.py_

Edit: moved `ShapeEnv` replay-recording to #107989

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ysiraichi committed:
Update on "Bisect FX node asserts on `ValidationException`."


This PR introduces binary search for finding smaller validation errors, when they occur.

We do that by bisecting the sequence of `torch._assert` FX nodes recorded as the source
expression of the translation validator (TV) by `ShapeEnv.evaluate_expr` calls. Then, we
raise the error caused by the earliest node.

In summary, the changes are:
- Call `bisect` on `ValidationError` @ _torch/_dynamo/convert_frame.py_
- Implement the binary search @ _torch/fx/experimental/symbolic_shapes.py_

Edit: moved `ShapeEnv` replay-recording to #107989

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ysiraichi committed:
Update on "Bisect FX node asserts on `ValidationException`."


This PR introduces binary search for finding smaller validation errors, when they occur.

We do that by bisecting the sequence of `torch._assert` FX nodes recorded as the source
expression of the translation validator (TV) by `ShapeEnv.evaluate_expr` calls. Then, we
raise the error caused by the earliest node.

In summary, the changes are:
- Call `bisect` on `ValidationError` @ _torch/_dynamo/convert_frame.py_
- Implement the binary search @ _torch/fx/experimental/symbolic_shapes.py_

Edit: moved `ShapeEnv` replay-recording to #107989

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ysiraichi committed:
Update on "Bisect FX node asserts on `ValidationException`."


This PR introduces binary search for finding smaller validation errors, when they occur.

We do that by bisecting the sequence of `torch._assert` FX nodes recorded as the source
expression of the translation validator (TV) by `ShapeEnv.evaluate_expr` calls. Then, we
raise the error caused by the earliest node.

In summary, the changes are:
- Call `bisect` on `ValidationError` @ _torch/_dynamo/convert_frame.py_
- Implement the binary search @ _torch/fx/experimental/symbolic_shapes.py_

Edit: moved `ShapeEnv` replay-recording to #107989

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ysiraichi committed:
Update on "Bisect FX node asserts on `ValidationException`."


This PR introduces binary search for finding smaller validation errors, when they occur.

We do that by bisecting the sequence of `torch._assert` FX nodes recorded as the source
expression of the translation validator (TV) by `ShapeEnv.evaluate_expr` calls. Then, we
raise the error caused by the earliest node.

In summary, the changes are:
- Call `bisect` on `ValidationError` @ _torch/_dynamo/convert_frame.py_
- Implement the binary search @ _torch/fx/experimental/symbolic_shapes.py_

Edit: moved `ShapeEnv` replay-recording to #107989

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ysiraichi committed:
Refactored into another file. on "Bisect FX node asserts on `ValidationException`."


This PR introduces binary search for finding smaller validation errors, when they occur.

We do that by bisecting the sequence of `torch._assert` FX nodes recorded as the source
expression of the translation validator (TV) by `ShapeEnv.evaluate_expr` calls. Then, we
raise the error caused by the earliest node.

In summary, the changes are:
- Call `bisect` on `ValidationError` @ _torch/_dynamo/convert_frame.py_
- Implement the binary search @ _torch/fx/experimental/symbolic_shapes.py_

Edit: moved `ShapeEnv` replay-recording to #107989

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ysiraichi committed:
Update on "Bisect FX node asserts on `ValidationException`."


This PR introduces binary search for finding smaller validation errors, when they occur.

We do that by bisecting the sequence of `torch._assert` FX nodes recorded as the source
expression of the translation validator (TV) by `ShapeEnv.evaluate_expr` calls. Then, we
raise the error caused by the earliest node.

In summary, the changes are:
- Call `bisect` on `ValidationError` @ _torch/_dynamo/convert_frame.py_
- Implement the binary search @ _torch/fx/experimental/symbolic_shapes.py_

Edit: moved `ShapeEnv` replay-recording to #107989

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ysiraichi committed:
Update on "Bisect FX node asserts on `ValidationException`."


This PR introduces binary search for finding smaller validation errors, when they occur.

We do that by bisecting the sequence of `torch._assert` FX nodes recorded as the source
expression of the translation validator (TV) by `ShapeEnv.evaluate_expr` calls. Then, we
raise the error caused by the earliest node.

In summary, the changes are:
- Call `bisect` on `ValidationError` @ _torch/_dynamo/convert_frame.py_
- Implement the binary search @ _torch/fx/experimental/symbolic_shapes.py_

Edit: moved `ShapeEnv` replay-recording to #107989

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ysiraichi committed:
Update on "Bisect FX node asserts on `ValidationException`."


This PR introduces binary search for finding smaller validation errors, when they occur.

We do that by bisecting the sequence of `torch._assert` FX nodes recorded as the source
expression of the translation validator (TV) by `ShapeEnv.evaluate_expr` calls. Then, we
raise the error caused by the earliest node.

In summary, the changes are:
- Call `bisect` on `ValidationError` @ _torch/_dynamo/convert_frame.py_
- Implement the binary search @ _torch/fx/experimental/symbolic_shapes.py_

Edit: moved `ShapeEnv` replay-recording to #107989

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ysiraichi committed:
Rebased. on "Bisect FX node asserts on `ValidationException`."


This PR introduces binary search for finding smaller validation errors, when they occur.

We do that by bisecting the sequence of `torch._assert` FX nodes recorded as the source
expression of the translation validator (TV) by `ShapeEnv.evaluate_expr` calls. Then, we
raise the error caused by the earliest node.

In summary, the changes are:
- Call `bisect` on `ValidationError` @ _torch/_dynamo/convert_frame.py_
- Implement the binary search @ _torch/fx/experimental/symbolic_shapes.py_

Edit: moved `ShapeEnv` replay-recording to #107989

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ysiraichi committed:
Fix imports. on "Bisect FX node asserts on `ValidationException`."


This PR introduces binary search for finding smaller validation errors, when they occur.

We do that by bisecting the sequence of `torch._assert` FX nodes recorded as the source
expression of the translation validator (TV) by `ShapeEnv.evaluate_expr` calls. Then, we
raise the error caused by the earliest node.

In summary, the changes are:
- Call `bisect` on `ValidationError` @ _torch/_dynamo/convert_frame.py_
- Implement the binary search @ _torch/fx/experimental/symbolic_shapes.py_

Edit: moved `ShapeEnv` replay-recording to #107989

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ysiraichi committed:
Rebased. on "Bisect FX node asserts on `ValidationException`."


This PR introduces binary search for finding smaller validation errors, when they occur.

We do that by bisecting the sequence of `torch._assert` FX nodes recorded as the source
expression of the translation validator (TV) by `ShapeEnv.evaluate_expr` calls. Then, we
raise the error caused by the earliest node.

In summary, the changes are:
- Call `bisect` on `ValidationError` @ _torch/_dynamo/convert_frame.py_
- Implement the binary search @ _torch/fx/experimental/symbolic_shapes.py_

Edit: moved `ShapeEnv` replay-recording to #107989

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ysiraichi committed:
Update on "Bisect FX node asserts on `ValidationException`."


This PR introduces binary search for finding smaller validation errors, when they occur.

We do that by bisecting the sequence of `torch._assert` FX nodes recorded as the source
expression of the translation validator (TV) by `ShapeEnv.evaluate_expr` calls. Then, we
raise the error caused by the earliest node.

In summary, the changes are:
- Call `bisect` on `ValidationError` @ _torch/_dynamo/convert_frame.py_
- Implement the binary search @ _torch/fx/experimental/symbolic_shapes.py_

Edit: moved `ShapeEnv` replay-recording to #107989

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ysiraichi committed:
Sort strigified Z3 expressions. on "Bisect FX node asserts on `ValidationException`."


This PR introduces binary search for finding smaller validation errors, when they occur.

We do that by bisecting the sequence of `torch._assert` FX nodes recorded as the source
expression of the translation validator (TV) by `ShapeEnv.evaluate_expr` calls. Then, we
raise the error caused by the earliest node.

In summary, the changes are:
- Call `bisect` on `ValidationError` @ _torch/_dynamo/convert_frame.py_
- Implement the binary search @ _torch/fx/experimental/symbolic_shapes.py_

Edit: moved `ShapeEnv` replay-recording to #107989

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ezyang commented on torch/fx/experimental/recording.py:
```diff
@@ -163,6 +163,13 @@ def is_evaluate_expr(self) -> bool:
     def is_defer_runtime_assert(self) -> bool:
         return self.name == "defer_runtime_assert"
 
+    def getarg(self, *, index: int, key: str) -> Any:
```
What does this do? Can we get some docs?

ysiraichi replied:
This is so we can get an argument from an event. It gets an argument by its positional index or its name.

ysiraichi replied:
I guess we can get rid of it, though. It's not a good solution for getting arguments from events.

ysiraichi committed:
Remove getarg function. on "Bisect FX node asserts on `ValidationException`."


This PR introduces binary search for finding smaller validation errors, when they occur.

We do that by bisecting the sequence of `torch._assert` FX nodes recorded as the source
expression of the translation validator (TV) by `ShapeEnv.evaluate_expr` calls. Then, we
raise the error caused by the earliest node.

In summary, the changes are:
- Call `bisect` on `ValidationError` @ _torch/_dynamo/convert_frame.py_
- Implement the binary search @ _torch/fx/experimental/symbolic_shapes.py_

Edit: moved `ShapeEnv` replay-recording to #107989

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2412,6 +2412,11 @@ def remove_fx_node(self, node: Optional[torch.fx.Node]) -> None:
 
+    def add_fx_node_metadata(self, node: torch.fx.Node) -> None:
+        from torch._dynamo.utils import get_current_node
+        node.meta["event"] = self.last_event_index()
```
This is too generic. Meta dict is used by everyone. Give a more detailed string key

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -3203,7 +3208,9 @@ def issue_guard(guard: ShapeGuard) -> None:
-            if self._maybe_evaluate_static(guard.expr) is not None:
+            static_expr = self._maybe_evaluate_static(guard.expr)
+            if static_expr is not None:
+                self.log.debug("eval %s == %s [statically known]", guard.expr, static_expr)
```
Are you sure about this log? We don't log other `issue_guard` here. This seems like it could be confuisng.

ezyang commented on torch/fx/experimental/validator.py:
```diff
@@ -559,18 +571,167 @@ def symbolstr(sym) -> str:
+
+    if not last_exception:
+        # We don't actually fail due to a produce_guards call.
+        # Stop and don't bisect.
```
maybe worth logging here?

ezyang commented on torch/fx/experimental/validator.py:
```diff
@@ -559,18 +571,167 @@ def symbolstr(sym) -> str:
+        f"bisecting expects {event.name} to have at least 2 positional arguments. "
+        f"Got: {len(args)}"
+    )
+    sympy.Eq
```
?

ysiraichi committed:
Update on "Bisect FX node asserts on `ValidationException`."


This PR introduces binary search for finding smaller validation errors, when they occur.

We do that by bisecting the sequence of `torch._assert` FX nodes recorded as the source
expression of the translation validator (TV) by `ShapeEnv.evaluate_expr` calls. Then, we
raise the error caused by the earliest node.

In summary, the changes are:
- Call `bisect` on `ValidationError` @ _torch/_dynamo/convert_frame.py_
- Implement the binary search @ _torch/fx/experimental/symbolic_shapes.py_

Edit: moved `ShapeEnv` replay-recording to #107989

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ysiraichi commented:
@pytorchbot merge

ezyang commented:
@pytorchbot merge -i
